perf/pkeyread.c

```diff
@@ -0,0 +1,312 @@
+     */
+    for (i = 0; i < NUM_CALLS_PER_TEST / threadcount; i++) {
+        key = PEM_read_bio_PrivateKey(pem, NULL, NULL, NULL);
+        if (key == NULL) {
+            fprintf(stderr, "Failed to create key: %d [%s PEM]\n", i,
+                    sample_names[sample_id]);
+            err = 1;
+            BIO_free(pem);
+            return;
+        }
+        EVP_PKEY_free(key);
+        BIO_reset(pem);
+    }
+
+    end = ossl_time_now();
+    times[num] = ossl_time_subtract(end, start);
+
+    BIO_free(pem);
+}
+
+void do_derread(size_t num)
+{
+    const char *keydata;
+    size_t keydata_sz;
+    const unsigned char *pk_buf = NULL;
+    int pk_buf_len;
+    BIO *der;
+    PKCS8_PRIV_KEY_INFO *pkey_info8 = NULL;
+    EVP_PKEY *pkey = NULL;
+    int i;
+    OSSL_TIME start, end;
+
+    if (sample_id >= SAMPLE_ALL) {
+        fprintf(stderr, "%s no sample key set for test\n", __func__);
+        err = 1;
+        return;
+    }
+
+    keydata = sample_keys[sample_id][FORMAT_DER];
+    keydata_sz = sample_key_sizes[sample_id][FORMAT_DER];
+    der = BIO_new_mem_buf(keydata, keydata_sz);
+    if (der == NULL) {
+        fprintf(stderr, "%s Cannot create mem BIO [%s DER]\n",
+                __func__, sample_names[sample_id]);
+        err = 1;
+        return;
+    }
+
+    start = ossl_time_now();
+
+    for (i = 0; i < NUM_CALLS_PER_TEST / threadcount && err == 0; i++) {
+        pkey_info8 = d2i_PKCS8_PRIV_KEY_INFO_bio(der, &pkey_info8);
+        if (pkey_info8 == NULL) {
+            fprintf(stderr, "%s pkey_info8 is NULL [%s BER]\n",
+                    __func__, sample_names[sample_id]);
+            err = 1;
+            goto error;
+        }
+
+        pkey = EVP_PKCS82PKEY(pkey_info8);
+        if (pkey == NULL) {
+            fprintf(stderr, "%s pkey is NULL [%s BER]\n",
+                    __func__, sample_names[sample_id]);
+            err = 1;
+            goto error;
+        }
+error:
+        BIO_reset(der);
+        EVP_PKEY_free(pkey);
+        pkey = NULL;
+        PKCS8_PRIV_KEY_INFO_free(pkey_info8);
+        pkey_info8 = NULL;
+    }
+
+    end = ossl_time_now();
+    times[num] = ossl_time_subtract(end, start);
+
+    BIO_free(der);
+}
+
+static int sample_name_to_id(const char *sample_name)
+{
+    int i = 0;
+
+    while (sample_names[i] != NULL) {
+        if (strcasecmp(sample_names[i], sample_name) == 0)
+            break;
+        i++;
+    }
+
+    return i;
+}
+
+static int format_name_to_id(const char *format_name)
+{
+    int i = 0;
+
+    while (format_names[i] != NULL) {
+        if (strcasecmp(format_names[i], format_name) == 0)
+            break;
+        i++;
+    }
+
+    return i;
+}
+
+double get_avcalltime(void)
+{
+    int i;
+    OSSL_TIME t;
+    double avcalltime;
+
+    for (i = 1; i < threadcount; i++)
+        t = ossl_time_add(t, times[i]);
+    avcalltime = (double)ossl_time2ticks(t) / (double)NUM_CALLS_PER_TEST;
+
+    avcalltime = (double)ossl_time2ticks(t) / (double)OSSL_TIME_US;
```

Comment:
This looks wrong. Shouldn't that be

```suggestion
    avcalltime = avcalltime / (double)OSSL_TIME_US;
```